src/app/MessageDef.h

```diff
@@ -0,0 +1,894 @@
+    EventDataElement::Builder SystemTimestamp(const uint64_t aSystemTimestamp);
+    EventDataElement::Builder DeltaUTCTime(const uint64_t aDeltaUTCTime);
+    EventDataElement::Builder DeltaSystemTime(const uint64_t aDeltaSystemTime);
+
+    // Mark the end of this array and recover the type for outer container
+    EventDataElement::Builder & EndOfEventDataElement();
+
+private:
+    EventPath::Builder mEventPathBuilder;
+};
+}; // namespace EventDataElement
+
+namespace EventList {
+class Parser : public ListParserBase
+{
+public:
+    // Roughly verify the schema is right, including
+    // 1) at least one element is there
+    // 2) all elements are anonymous and of Structure type
+    // 3) every Event is also valid in schema
+    CHIP_ERROR CheckSchemaValidity() const;
+};
+
+class Builder : public ListBuilderBase
+{
+public:
+    // Re-initialize the shared EventBuilder with anonymous tag
+    EventDataElement::Builder & CreateEventBuilder();
+
+    // Mark the end of this array and recover the type for outer container
+    EventList::Builder & EndOfEventList();
+
+private:
+    EventDataElement::Builder mEventDataElementBuilder;
+};
+}; // namespace EventList
+
+namespace StatusElement {
+enum
+{
+    kCsTag_GeneralCode         = 1,
+    kCsTag_ProtocolId          = 2,
+    kCsTag_ProtocolCode        = 3,
+    kCsTag_NamespacedClusterId = 4
+};
+
+class Parser : public ListParserBase
+{
+public:
+    // aReader has to be on the element of DataElement
+    CHIP_ERROR Init(const chip::TLV::TLVReader & aReader);
+
+    // Roughly verify the schema is right, including
+    // 1) all mandatory tags are present
+    // 2) all elements have expected data type
+    // 3) any tag can only appear once
+    // At the top level of the structure, unknown tags are ignored for foward compatibility
+    CHIP_ERROR CheckSchemaValidity() const;
+
+    /**
+    `* Read the GeneralCode, ProtocolId, ProtocolCode, NamespacedClusterId
+     *
+     * @param[out]   apGeneralCode     Pointer to the storage for the GeneralCode
+     * @param[out]   apProtocolId     Pointer to the storage for the ProtocolId
+     * @param[out]   apProtocolCode   Pointer to the storage for the ProtocolCode
+     * @param[out]   apNamespacedClusterId     Pointer to the storage for the NamespacedClusterId
+     *
+     * @return       CHIP_ERROR codes returned by Chip::TLV objects. CHIP_END_OF_TLV if either
+     *               element is missing. CHIP_ERROR_WRONG_TLV_TYPE if the elements are of the wrong
+     *               type.
+     */
+    CHIP_ERROR DecodeStatusElement(uint16_t * apGeneralCode, uint32_t * apProtocolId, uint16_t * apProtocolCode,
+                                   chip::ClusterId * apNamespacedClusterId) const;
+};
+
+class Builder : public ListBuilderBase
+{
+public:
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter);
+    CHIP_ERROR Init(chip::TLV::TLVWriter * const apWriter, const uint8_t aContextTagToUse);
```

Comment:
the ListBuilderBase could have this other form of `Init()`

Reply:
may i get the clarification on this question?

Reply:
AttributePath, EventPath, CommandPath, they all inherit from BuildBase, in init, they start the container with kTLVType_Path. StartContainer(aTag, chip::TLV::kTLVType_Path, mOuterContainerType)
class ListBuilderBase(inherited by AttributePathList..) also inherits from BuilderBase, but it start the container with kTLVType_Array
mpWriter->StartContainer(chip::TLV::AnonymousTag, chip::TLV::kTLVType_Array, mOuterContainerType);
therefore, I am not putting the init and init with tag in BuildBase